snippets/titration/titration.py

```diff
@@ -0,0 +1,947 @@
+    intermediate_cells[i] = ad_subset[ad_subset.obs[method_column] == 'Inter'].obs['object_id']
+    negative_cells[i] = ad_subset[ad_subset.obs[method_column] == 'Negative'].obs['object_id']
+    
+    # Mask the cells
+    pos = np.array(positive_cells[i]).astype('int')
+    inter = np.array(intermediate_cells[i]).astype('int')
+    neg = np.array(negative_cells[i]).astype('int')
+
+    mask_overlays[i] = np.empty(shape(mask_list[i]), dtype=int)
+    mask_overlays[i].fill(0)
+    mask_overlays[i][np.isin(mask_list[i], pos)] = 3
+    mask_overlays[i][np.isin(mask_list[i], inter)] = 2
+    mask_overlays[i][np.isin(mask_list[i], neg)] = 1
+
+# %% [markdown]
+# ### Display images and masks
+#
+# __Row 1:__ Selected images, the current marker is shown.
+#
+# __Row 2:__ Masks plotted onto images.
+# Positive cells are overlaid in blue, negative cells in red and intermediate cells in yellow.
+#
+# ___Manually adjust image brightness and mask transparency if needed___
+
+# %%
+# Brightness and transparency
+brightness = 20
+transparency = 0.3
+
+# Display images
+fig, axs = plt.subplots(2, len(concentrations), figsize=(25, 25))
+
+for i, conc in enumerate(concentrations):
+    # Images
+    cur_image = cv2.convertScaleAbs(image_list[i], alpha=brightness, beta=0)
+    cur_image = cur_image[crop_mask[i][0]:crop_mask[i][1], crop_mask[i][2]:crop_mask[i][3]]
+    axs[0,i].imshow(cur_image, cmap=plt.cm.viridis)
+    axs[0,i].set_title("Concentration =" + str(conc))
+    axs[0,i].axis('off')
+
+    # Image-Mask overlay
+    overlay = cur_image.copy()
+    overlay = skimage.color.gray2rgb(cur_image)
+    cur_mask = mask_overlays[i]
+    cur_mask = cur_mask[crop_mask[i][0]:crop_mask[i][1], crop_mask[i][2]:crop_mask[i][3]]
+    overlay = skimage.color.label2rgb(cur_mask, overlay, alpha=transparency, bg_label=0,
+                                     colors=['blue','yellow','red'])
+    axs[1,i].imshow((overlay*255).astype('uint8'))
+    axs[1,i].set_title("Concentration =" + str(conc))
+    axs[1,i].axis('off')
+    
+plt.show()
+
+# %% [markdown]
+# ### View images and masks with napari
+# There is also the option to view the images with napari (no cropping and interactive zoom / mask overlay).  
+#
+# First, select the concentration you want to see:
+
+# %%
+# w_conc = widgets.Select(
+#     options=concentrations,
+#     value=concentrations[0],
+#     description='Concentrations:',
+#     disabled=False
+# )
+# w_conc
+
+# %%
+# cur_conc = np.where(concentrations == w_conc.value)
+# print("Current concentration = ", double(concentrations[cur_conc]))
+
+# %% [markdown]
+# __Start napari viewer__  
+# Positive cells are shown in white and negative cells in black on the mask
+
+# %%
+# viewer = napari.Viewer() 
+# viewer.add_image(image_list[int(cur_conc[0])], name = 'IMC image ' + cur_marker)
+# viewer.add_image(mask_overlays[int(cur_conc[0])], name = 'Cell mask')
+# napari.run()
```

Comment:
Maybe this is of interest to you, primarily for the part about loading multi-channel images: https://gist.github.com/jwindhager/db0d13cb856948b072c91514e1586e0b

Reply:
Nice! I will add it in a next version, napari is not a priority.
Is it normal that the "outline" image doesn't show the outline of every single cell?